### PR TITLE
Test: Bind delta merge storage with executor ut

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609

Problem Summary:
We want to use `DeltaMergeStorage` to run the query in `ut framework`.
Then we can test the whole query process in single node(not including mpp with multiple nodes, we assume that data are stored in one node).
It will be helpful for features that change both `storage code` and `compute code`. Eg: late materialization and pipeline model.
Maybe RSFilter can use `ut framework` too.
Here is one simple case:
```c++
context.addMockDeltaMerge({"test_db", "t0"},
                                  {{"col0", TiDB::TP::TypeLongLong}},
                                  {{toVec<Int64>("col0", {0, 1, 2, 3, 4, 5, 6, 7})}});

auto request = context
                       .scan("test_db", "t0")
                       .build(context);
executeAndAssertColumnsEqual(
        request,
        {{toNullableVec<Int64>("col0", {0, 1, 2, 3, 4, 5, 6, 7})}});
```
### What is changed and how it works?
1. We use `MockStorage` to hold `DeltaMergeStorage`.
```c++
    std::unordered_map<String, Int64> name_to_id_map_for_delta_merge; /// <table_name, table_id>
    std::unordered_map<Int64, MockColumnInfoVec> table_schema_for_delta_merge; /// <table_id, columnInfo>
    std::unordered_map<Int64, std::shared_ptr<StorageDeltaMerge>> storage_delta_merge_map; // <table_id, StorageDeltaMerge>
    std::unordered_map<String, TableInfo> table_infos_for_delta_merge; /// <table_name, table_info>
    std::unordered_map<Int64, NamesAndTypes> names_and_types_map_for_delta_merge; /// <table_id, NamesAndTypes>
```
2. When writing the test, we must call 
```c++
MockDAGRequestContext::addMockDeltaMerge(const MockTableName & name, const MockColumnInfoVec & columnInfos, ColumnsWithTypeAndName columns);
```
to define the table schema and insert table columns.
3. `addMockDeltaMerge` will call
```c++
addMockDeltaMergeSchema(db, table, columnInfos);
addMockDeltaMergeData(db, table, columns);
```
to add schema and data into corresponding `DeltaMergeStorage`
4. When running a test query, planner will choose to use the `DeltaMergeStorage` to generate source streams.
5. BTW, I have moved a bunch of code like
```c++
using ColMyDateTimeNullableType = std::optional<typename TypeTraits<MyDateTime>::FieldType>;
using ColDecimalNullableType = std::optional<typename TypeTraits<Decimal32>::FieldType>;
```
into `dbms/src/TestUtils/ExecutorTestUtils.h` to make the test code cleaner.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Debug/MockStorage.cpp [hunk 1]
```diff
@@ -63,6 +72,143 @@ MockColumnInfoVec MockStorage::getTableSchema(const String & name)
...
+
+        output->writePrefix();
+        output->write(insert_block);
+        output->writeSuffix();
+    }
+}
+
+BlockInputStreamPtr MockStorage::getStreamFromDeltaMerge(Context & context, Int64 id)
+{
+    auto storage = storage_delta_merge_map[id];
```
Comment (SeaRise) [inline dbms/src/Debug/MockStorage.cpp:124]: add some assert for function `xxxFromDeltaMerge`

File: dbms/src/Flash/tests/gtest_executors_with_dm.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                                   toNullableVec<Int8>("col1", col_tinyint),
+                                   toNullableVec<Int16>("col2", col_smallint),
+                                   toNullableVec<Int32>("col3", col_int),
+                                   toNullableVec<Int64>("col4", col_bigint),
+                                   toNullableVec<Float32>("col5", col_float),
+                                   toNullableVec<Float64>("col6", col_double),
+                                   toNullableVec<MyDate>("col7", col_mydate),
+                                   toNullableVec<MyDateTime>("col8", col_mydatetime),
+                                   toNullableVec<String>("col9", col_string)});
+        context.mockStorage()->setUseDeltaMerge();
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_executors_with_dm.cpp:62]: how about moving this line before `addMockDeltaMerge`.

**Review by SeaRise — APPROVED** [review]
Others lgtm

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**